hibernate-core/src/main/java/org/hibernate/loader/ast/internal/ExecutionContextWithSubselectFetchHandler.java

```diff
@@ -4,30 +4,49 @@
  */
 package org.hibernate.loader.ast.internal;
 
+import org.hibernate.LockOptions;
 import org.hibernate.engine.spi.EntityHolder;
 import org.hibernate.engine.spi.SharedSessionContractImplementor;
 import org.hibernate.engine.spi.SubselectFetch;
 import org.hibernate.query.spi.QueryOptions;
+import org.hibernate.query.spi.QueryOptionsAdapter;
 import org.hibernate.sql.exec.internal.BaseExecutionContext;
 
 class ExecutionContextWithSubselectFetchHandler extends BaseExecutionContext {
 
 	private final SubselectFetch.RegistrationHandler subSelectFetchableKeysHandler;
 	private final boolean readOnly;
+	private final QueryOptions queryOptions;
 
 	public ExecutionContextWithSubselectFetchHandler(
 			SharedSessionContractImplementor session,
 			SubselectFetch.RegistrationHandler subSelectFetchableKeysHandler) {
-		this( session, subSelectFetchableKeysHandler, false );
+		this( session, subSelectFetchableKeysHandler, false, null );
 	}
 
 	public ExecutionContextWithSubselectFetchHandler(
 			SharedSessionContractImplementor session,
 			SubselectFetch.RegistrationHandler subSelectFetchableKeysHandler,
-			boolean readOnly) {
+			boolean readOnly,
+			LockOptions lockOptions) {
 		super( session );
 		this.subSelectFetchableKeysHandler = subSelectFetchableKeysHandler;
 		this.readOnly = readOnly;
+		this.queryOptions = determineQueryOptions( readOnly, lockOptions );
+	}
+
+	private QueryOptions determineQueryOptions(boolean readOnly, LockOptions lockOptions) {
+		return new QueryOptionsAdapter() {
+			@Override
+			public Boolean isReadOnly() {
+				return readOnly;
+			}
+
+			@Override
+			public LockOptions getLockOptions() {
+				return lockOptions == null ? LockOptions.NONE : lockOptions;
+			}
+		};
```

Comment:
We can use `SimpleQueryOptions` here, just to avoid unnecessary anonymous inner types. Also, `isReadOnly` returning `false` has different semantics that returning `null` (which is what happened before, see `QueryOptions.NONE`) so we should preserve that behavior:
```suggestion
		return new new SimpleQueryOptions( lockOptions, readOnly ? true : null );
```